### PR TITLE
Implemented mode function

### PR DESCRIPTION
**Pull request by Yash621**
this PR is respect to #8978 , @jakevdp  please review it.

I have implemented the mode function ,and have also created the _core.py file  and have placed the function there.
also i have added the license required and written unit tests as you guided under scipy_stats_test.py file.

please review my PR and let me know if any changes are required.

### CODE REVIEW

File: jax/_src/api.py [hunk 1]
```diff
@@ -2547,8 +2547,8 @@ def linear_transpose(fun: Callable, *primals, reduce_axes=()) -> Callable:
                                                instantiate=True)
   out_avals, _ = unzip2(out_pvals)
   out_dtypes = map(dtypes.dtype, out_avals)
-  if not (all(dtypes.issubdtype(d, np.inexact) for d in in_dtypes + out_dtypes)
-          or all(dtypes.issubdtype(d, np.integer)
+  if not (all(dtypes.issubdtype(d, np.inexact) for d in in_dtypes + out_dtypes) or
+          all(dtypes.issubdtype(d, np.integer)
                  for d in in_dtypes + out_dtypes)):
     raise TypeError("linear_transpose only supports [float or complex] -> "
                     "[float or complex], and integer -> integer functions, "
```
Comment (jakevdp) [inline jax/_src/api.py:2554]: Please revert changes to this file; they seem unrelated to your change.

File: jax/_src/scipy/stats/_core.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    ...               [5, 3, 0, 5],
+    ...               [4, 7, 5, 9]])
+    >>> from scipy import stats
+    >>> stats.mode(a)
+    ModeResult(mode=array([[3, 1, 0, 0]]), count=array([[1, 1, 1, 1]]))
+    To get mode of whole array, specify ``axis=None``:
+    >>> stats.mode(a, axis=None)
+    ModeResult(mode=array([3]), count=array([3]))
+    """
+
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:105]: We should call `_check_arraylike()` here to ensure that the input is not a list (search through the codebase to see how this function is used elsewhere)

File: jax/_src/scipy/stats/_core.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    a, axis = _chk_asarray(a, axis)
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:106]: I think it would be clearer to put the contents of `_chk_asarray` inline – it's only a few lines of code, and is not used elsewhere, so there's not much benefit to wrapping it in a function.

File: jax/_src/scipy/stats/_core.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            warnings.warn("The input array could not be properly "
+                          "checked for nan values. nan values "
+                          "will be ignored.", RuntimeWarning)
+
+    if contains_nan and nan_policy == 'raise':
+        raise ValueError("The input contains nan values")
+
+    return contains_nan, nan_policy
+
+def mode(a, axis=0, nan_policy='propagate'):
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:66]: Rather than copying the contents of the docstring from scipy (which has licensing implications), we instead use the `_wraps` decorator to add it automatically. You can look at functions defined in `jax/_src/numpy/lax_numpy.py` to see how it is used.

File: jax/_src/scipy/stats/_core.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    To get mode of whole array, specify ``axis=None``:
+    >>> stats.mode(a, axis=None)
+    ModeResult(mode=array([3]), count=array([3]))
+    """
+
+    a, axis = _chk_asarray(a, axis)
+    if a.size == 0:
+        return ModeResult(jnp.array([]), jnp.array([]))
+
+    contains_nan, nan_policy = _contains_nan(a, nan_policy)
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:110]: `nan_policy = 'raise'` will never work correctly in traced code; we handle this elsewhere by detecting this and raising an appropriate error. See for example the handling of `mode='raise'` here: https://github.com/google/jax/blob/1cf7d4ab5dcea4619a2b6c1b471f2be340b6abfc/jax/_src/numpy/lax_numpy.py#L3462

File: jax/_src/scipy/stats/_core.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    if contains_nan and nan_policy == 'omit':
+        a = ma.masked_invalid(a)
+        return mstats_basic.mode(a, axis)
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:114]: We cannot fall back to numpy/scipy functions here, because they cast all inputs to regular numpy arrays, and thus are incompatible with JAX transforms. We'll have to implement these directly (or, you can punt on it now and raise `NotImplementedError`)

File: jax/_src/scipy/stats/_core.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+    if a.dtype == object and jnp.nan in set(a.ravel()):
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:116]: dtype is never object in JAX. So perhaps this whole block should be removed, because it seems to handle a case that will never occur?

File: tests/lax_numpy_test.py [hunk 8]
```diff
@@ -6070,8 +6069,8 @@ class NumpyUfuncTests(jtu.JaxTestCase):
   def testUfuncInputTypes(self, name, arg_dtypes):
     # TODO(jakevdp): fix following failures and remove from this exception list.
     if (name in ['divmod', 'floor_divide', 'fmod', 'gcd', 'left_shift', 'mod',
-                 'power', 'remainder', 'right_shift', 'rint', 'square']
-        and 'bool_' in arg_dtypes):
+                 'power', 'remainder', 'right_shift', 'rint', 'square'] and
+        'bool_' in arg_dtypes):
       self.skipTest(f"jax.numpy does not support {name}{tuple(arg_dtypes)}")
     if name == 'arctanh' and jnp.issubdtype(arg_dtypes[0], jnp.complexfloating):
       self.skipTest("np.arctanh & jnp.arctanh have mismatched NaNs for complex input.")
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:6076]: Please revert unrelated formatting changes.

File: jax/_src/scipy/stats/_core.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    # This recreates the results without that issue
+    # View of a, rotated so the requested axis is last
+    in_dims = list(range(a.ndim))
+    a_view = jnp.transpose(a, in_dims[:axis] + in_dims[axis+1:] + [axis])
+
+    inds = np.ndindex(a_view.shape[:-1])                                    #star
+    modes = jnp.empty(a_view.shape[:-1], dtype=a.dtype)
+    counts = jnp.empty(a_view.shape[:-1], dtype=jnp.int_)
+    for ind in inds:
+        modes[ind], counts[ind] = _mode1D(a_view[ind])
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:148]: JAX arrays are immutable, so you cannot assign their values this way.

But it looks like this is an attempt to map a 1D function to multiple dimensions... for performance reasons, we should lean toward doing this with `vmap` rather than by manual iteration.

File: tests/scipy_stats_test.py [hunk 10]
```diff
@@ -623,6 +623,38 @@ def testMultivariateNormalLogpdfBatch(self, ndim, nbatch, dtype):
...
+         "axis": axis,
+         "keepdims": keepdims}
+        for a_dtype in jtu.dtypes.floating
+        for a_shape, axis in (
+          ((7,), None),
+          ((47, 7), 0),
+          ((4, 101), 1),
+        )
+        for keepdims in [False, True]
+        for op in ["mode", "nanmode"]))
```
Comment (jakevdp) [inline tests/scipy_stats_test.py:640]: I'm confused by the use of `nanmode` here. In similar tests within `lax_numpy_test.py`, we are testing the functions e.g. `jnp.mean` and `jnp.nanmean`. I don't think there's a `nanmode` function though?

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks for pushing on this! I have a number of comments – in particular we need to be careful about handling traced quantities correctly (This might be some good background reading: https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html)

**Comment by jakevdp** [comment]
Also, please do not close this PR and open another one. You can make the requested changes by pushing to your existing branch – it's far easier to track the review process in a single PR.

**Comment by Yash621** [comment]
@jakevdp  thanks for the reviews ,I will keep that in mind and implement the changes required

**Comment by Yash621** [comment]
@jakevdp  I have implemented all the changes suggested ,please review them.
and let me know if any further changes are required :)

**Comment by jakevdp** [comment]
There are still a number of formatting changes unrelated to the `mode` function. Please revert those.

**Comment by jakevdp** [comment]
I would suggest running the tests locally – the CI failures suggest that there may be some syntax or import issues.

**Comment by Yash621** [comment]
@jakevdp  okay !
 How can i run them locally ?

**Comment by jakevdp** [comment]
See https://jax.readthedocs.io/en/latest/developer.html#running-the-tests

**Comment by Yash621** [comment]
@jakevdp do we need to import the function in jax/scipy/stats/__init__.py or jax/_src/scipy/stats/__init__.py ?

**Comment by jakevdp** [comment]
Do you have un-committed changes locally? Because there are undefined variables in the versions of the files that you pushed to this branch.

**Comment by Yash621** [comment]
@jakevdp I have fixed undefined variables .
please review
and let me know if any further changes are required :)

**Comment by jakevdp** [comment]
Your PR still contains a large number of unrelated changes, some of which are leading to lint errors (such as adding stray spaces at the end of a number of lines). I've mentioned this a few times already, but please revert these changes.

**Comment by jakevdp** [comment]
We don't use any automated formatter, unfortunately.

**Comment by Yash621** [comment]
@jakevdp  I have reverted all the unrelated changes manually.
please review
and let me know if any further changes are required :)

**Comment by jakevdp** [comment]
There are still a number of unnecessary whitespace changes that are leading to lint errors, for example in `jax/_src/numpy/lax_numpy.py` and `jax/_src/api.py`. Please revert these as well.

**Comment by Yash621** [comment]
@jakevdp I have reverted the blank lines but i couldn't found any trailing whitespaces ,in the files
I have compared them from the original JAX repo.

**Comment by jakevdp** [comment]
Click on the "Files changed" tab above and you'll see the whitespace diffs. If you can't see those diffs locally, it might be that you have some kind of git hook set up that re-formats your changes when you commit them.

**Comment by Yash621** [comment]
@jakevdp  I have fixed all the whitespaces issues.
please review
and let me know if any further changes are required :)

**Comment by jakevdp** [comment]
Please make sure the lint & typecheck tests pass. You can run them locally following the instructions here: https://jax.readthedocs.io/en/latest/contributing.html#linting-and-type-checking

**Comment by Yash621** [comment]
@jakevdp I have fixed the linting errors ,but the type errors are in other files that i have not worked on.

**Comment by jakevdp** [comment]
The test you added is still failing. I'd suggest running it locally as you debug – you should make certain the test passes before pushing the updates here.

**Comment by Yash621** [comment]
@jakevdp I am getting assertion error on running the tests 

  AssertionError: Tuples differ: (1,) != ()
E   
E   First tuple contains 1 additional elements.
E   First extra element 0:
E   1
E   
E   - (1,)
E   + ()

How can I track for which testcase it is throwing this error?
and is it returning the shape or actual output of the result ?

**Comment by jakevdp** [comment]
The full test output should contain the name of the test that is failing.

**Comment by jakevdp** [comment]
Thanks for the work on this – closing in favor of #12683
